File: app/src/main/java/com/exzell/mangaplayground/viewmodels/SearchViewModel.java
package com.exzell.mangaplayground.viewmodels;

import android.app.Application;
import android.content.Context;
import android.os.Handler;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.SavedStateHandle;

import com.exzell.mangaplayground.adapters.MangaListAdapter;
import com.exzell.mangaplayground.advancedsearch.MangaSearch;
import com.exzell.mangaplayground.io.Repository;
import com.exzell.mangaplayground.models.Manga;
import com.exzell.mangaplayground.utils.MangaUtils;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchViewModel extends AndroidViewModel {
    private final String TAG = "SearchViewModel";
    private Context mContext;
    private Repository mRepo;
    private SavedStateHandle mHandle;
    private Handler mHandler;

    //KEYS
    private final String KEY_TITLE = "title";
    private final String KEY_AUTHOR = "auth/art";
    private final String KEY_TITLE_CONTAIN = "title_extra";
    private final String KEY_AUTHOR_CONTAIN = "auth/art_extra";
    private final String KEY_GENRE = "genres";
    private final String KEY_RATING = "getRating";
    private final String KEY_STATUS = "status";
    private final String KEY_TYPE = "type";
    private final String KEY_CHAPTERS = "chapters";
    private final String KEY_RELEASE = "releases";
    private final String KEY_GENRE_INCL = "genre inclusion";
    private final String KEY_ORDER = "order";

    private final String KEY_LINK = "last search link";


    public static final List<String> statusData = Stream.of(MangaSearch.STATUS_COMPLETED, MangaSearch.STATUS_ONGOING).map(String::toUpperCase).collect(Collectors.toList());
    public static final List<Integer> chapterData = Stream.of(1, 5, 10, 20, 30, 40, 50, 100, 200).collect(Collectors.toList());
    public static final List<String> releaseData = IntStream.rangeClosed(1946, 2017).boxed().map(String::valueOf).collect(Collectors.toList());

    private List<Manga> mCurrentSearchResults = new ArrayList<>();

    public SearchViewModel(@NonNull Application application, SavedStateHandle handle) {
        super(application);
        mContext = application.getApplicationContext();
        mRepo = Repository.getInstance(application);
        mHandle = handle;
        mHandler = new Handler();
    }

    /**
     * Collects a map of search parameters to their values which can be created through {@link #search}
     * and makes the request as well as creating the mangas
     * The consumer is called on the same thread the request was made
     * @param onMangaRetrieved A consumer called when the mangas have been created
     */
    public void resolveSearch(Map<String, String> search, Consumer<List<Manga>> onMangaRetrieved){

            mRepo.advancedSearch(search).enqueue(new Callback<ResponseBody>() {
                @Override
                public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                    if(response.isSuccessful()) {
                        Element html;
                        try {
                            html = Jsoup.parse(response.body().string()).body();
                            response.body().close();

                            String next = html.getElementsContainingText("next▶").attr("href");

                            if (!next.isEmpty()) {

                                Map<String, String> nextLink = new HashMap<>(search);
                                nextLink.put("page", getDigits(next));
                                setNextSearchLink(nextLink);
                            }else setNextSearchLink(null);

                            List<Manga> searchManga = MangaUtils.createSearchManga(html);
                            mCurrentSearchResults.addAll(searchManga);
                            onMangaRetrieved.accept(searchManga);
                            Log.w(TAG, next);

                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }else{
                        Toast.makeText(mContext, "Error Code: " + response.code(), Toast.LENGTH_SHORT).show();
                    }
                }

                @Override
                public void onFailure(Call<ResponseBody> call, Throwable t) {
                    Toast.makeText(mContext,"Failed: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
    }

    public List<Manga> getCurrentSearchResults(){return mCurrentSearchResults;}

    public void clearSearchResults(){mCurrentSearchResults.clear();}

    public void setNextSearchLink(Map<String, String> nextLink){
        mHandle.set(KEY_LINK, nextLink);
    }

    public Map<String, String> getNextLink(){return mHandle.get(KEY_LINK);}

    private String getDigits(String string){
        StringBuilder num = new StringBuilder();
        for(char c : string.toCharArray()){
            if(Character.isDigit(c)) num.append(c);
        }

        return num.toString();
    }

    public List<String> getSelectedGenres(){
        return mHandle.get(KEY_GENRE);
    }

    public String getName(boolean which){
        return which ? mHandle.get(KEY_AUTHOR) : mHandle.get(KEY_TITLE);
    }

    public String containValue(boolean which){
        return which ? mHandle.get(KEY_AUTHOR_CONTAIN) : mHandle.get(KEY_TITLE_CONTAIN);
    }

    public int getRating(){return mHandle.get(KEY_RATING);}

    public String getStatus(){
        return mHandle.get(KEY_STATUS);
    }

    public String getType(){
        return mHandle.get(KEY_TYPE);
    }

    public int getChapters(){
        return mHandle.get(KEY_CHAPTERS);
    }

    public int getRelease(){return mHandle.get(KEY_RELEASE);}

    public String getGenreInclusion(){return mHandle.get(KEY_GENRE_INCL);}

    public String getOrder(){return mHandle.get(KEY_ORDER);}

    public void setGenres(String genre, boolean which){
        ArrayList<String> arr = mHandle.get(KEY_GENRE);
        if(which) arr.add(genre);
        else{
            arr.remove(arr.indexOf(genre));
        }
    }

    public void setName(boolean which, String value){
        String key =  which ? KEY_AUTHOR : KEY_TITLE;
        mHandle.set(key, value);
    }

    public void setContainValue(boolean which, String value){
        String key =  which ? KEY_AUTHOR_CONTAIN : KEY_TITLE_CONTAIN;
        mHandle.set(key, value);
    }

    public void setRating(int rating){ mHandle.set(KEY_RATING, rating);}

    public void setStatus(String status){ mHandle.set(KEY_STATUS, status); }

    public void setType(String type){
        mHandle.set(KEY_TYPE, type);
    }

    public void setChapters(int chap){
        mHandle.set(KEY_CHAPTERS, chap);
    }

    public void setRelease(int release){mHandle.set(KEY_RELEASE, release);}

    public void setGenreInclusion(String incl){mHandle.set(KEY_GENRE_INCL, incl);}

    public void setOrder(String order){
        mHandle.set(KEY_ORDER, order);
    }

    public void resetValues(){
        setName(false, "");
        setName(true, "");
        setContainValue(false, "");
        setContainValue(true, "");
        setChapters(-1);
        setRating(-1);
        setRelease(-1);
        setStatus("");
        setType("");
        setGenreInclusion("");
        getSelectedGenres().clear();
    }

    public void handlerDefaults(){
        if(!mHandle.contains(KEY_GENRE)) mHandle.set(KEY_GENRE, new ArrayList<String>());
        if(!mHandle.contains(KEY_TITLE)) mHandle.set(KEY_TITLE, "");
        if(!mHandle.contains(KEY_AUTHOR)) mHandle.set(KEY_AUTHOR, "");
        if(!mHandle.contains(KEY_TITLE_CONTAIN)) mHandle.set(KEY_TITLE_CONTAIN, "");
        if(!mHandle.contains(KEY_AUTHOR_CONTAIN)) mHandle.set(KEY_AUTHOR_CONTAIN, "");
        if(!mHandle.contains(KEY_RATING)) mHandle.set(KEY_RATING, -1);
        if(!mHandle.contains(KEY_STATUS)) mHandle.set(KEY_STATUS, "");
        if(!mHandle.contains(KEY_TYPE)) mHandle.set(KEY_TYPE, "");
        if(!mHandle.contains(KEY_CHAPTERS)) mHandle.set(KEY_CHAPTERS, -1);
        if(!mHandle.contains(KEY_RELEASE)) mHandle.set(KEY_RELEASE, -1);
        if(!mHandle.contains(KEY_GENRE_INCL)) mHandle.set(KEY_GENRE_INCL, "");
        if(!mHandle.contains(KEY_ORDER)) mHandle.set(KEY_ORDER, "");
    }

    public Map<String, String> search() {
        MangaSearch search = new MangaSearch.Builder().setAuthor(getName(true))
                .setAuthorContain(containValue(true))
                .setTitle(getName(false))
                .setTitleContain(containValue(false))
                .setChapters(getChapters())
                .setRating(getRating())
                .setRelease(getRelease())
                .setStatus(getStatus())
                .setType(getType())
                .addGenres(getSelectedGenres())
                .setGenreInclusion(getGenreInclusion())
                .setOrder(getOrder())
                .build();

//        if(search.searchQuery().isEmpty()) return;
        return search.searchQuery();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        mCurrentSearchResults.clear();
    }
}

File: app/src/main/java/com/exzell/mangaplayground/viewmodels/SearchViewModel.kt
package com.exzell.mangaplayground.viewmodels

import android.annotation.SuppressLint
import android.app.Application
import androidx.lifecycle.SavedStateHandle
import com.exzell.mangaplayground.advancedsearch.Genre
import com.exzell.mangaplayground.advancedsearch.MangaSearch
import com.exzell.mangaplayground.advancedsearch.Order
import com.exzell.mangaplayground.advancedsearch.Type
import com.exzell.mangaplayground.io.Repository
import com.exzell.mangaplayground.models.Manga
import com.exzell.mangaplayground.utils.createSearchManga
import com.exzell.mangaplayground.utils.isConnectedToNetwork
import com.exzell.mangaplayground.utils.toManga
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import org.jsoup.Jsoup
import java.util.*
import java.util.function.Consumer
import javax.inject.Inject
import kotlin.collections.HashMap

@SuppressLint("staticFieldLeak")
class SearchViewModel(application: Application, private val mHandle: SavedStateHandle) : DisposableViewModel(application) {

    companion object {
        const val ERROR_NO_RESULT = "error no result"
        const val ERROR_OTHERS = "error others"
    }

    private val mContext = application.applicationContext

    lateinit var mSearch: MangaSearch

    @Inject
    lateinit var mRepo: Repository

    private val KEY_LINK = "last search link"

    @JvmField
    val mCurrentSearchResults = ArrayList<Manga>()

    val nextLink: Map<String, String>?
        get() = mHandle.get<Map<String, String>>(KEY_LINK)

    var rating: Int
        get() = mSearch.rating
        set(rating) {
            mSearch.rating = rating
        }

    var status: String?
        get() = mSearch.status
        set(status) {
            mSearch.status = status
        }

    var type: String?
        get() = mSearch.type.dispName
        set(type) {
            mSearch.type = Type.values().find { it.dispName == type }
        }

    var chapters: Int
        get() = mSearch.chapterAmount
        set(chap) {
            mSearch.chapterAmount = chap
        }

    var release: Int
        get() = mSearch.release
        set(release) {
            mSearch.release = release
        }

    var genreInclusion: String?
        get() = mSearch.genreInclusion
        set(incl) {
            mSearch.genreInclusion = incl
        }

    var order: String?
        get() = mSearch.order.dispName
        set(order) {
            mSearch.order = Order.values().find { it.dispName == order }
        }

    var authorContain: String
        get() = mSearch.authorContain
        set(contain) {
            mSearch.authorContain = contain
        }

    var titleContain: String
        get() = mSearch.titleContain
        set(contain) {
            mSearch.titleContain = contain
        }

    var title: String
        get() = mSearch.title
        set(title) {
            mSearch.title = title
        }

    var author: String
        get() = mSearch.author
        set(author) {
            mSearch.author = author
        }

    var onSuccess: Consumer<List<Manga>>? = null

    var onError: Consumer<String>? = null

    /**
     * Collects a map of search parameters to their values which can be created through [.search]
     * and makes the request as well as creating the mangas
     */
    fun resolveSearch(search: Map<String, String>) {
        if (!mContext.isConnectedToNetwork()) {
            onError?.accept(ERROR_OTHERS)
            return
        }

        addDisposable(mRepo.advancedSearch(search)?.map {
            val html = Jsoup.parse(it.string()).body()
            it.close()

            val hasNoMatch = html.select("div[class=manga-list").hasClass("no-match")
            if (hasNoMatch) throw RuntimeException(ERROR_NO_RESULT)

            val next = html.getElementsContainingText("next▶").attr("href")

            if (next.isNotEmpty()) setNextSearchLink(next, search)
            else setNextSearchLink(null, search)

            html.createSearchManga(mCurrentSearchResults.size)
        }?.observeOn(AndroidSchedulers.mainThread())?.subscribe({
            mCurrentSearchResults.addAll(it)
            onSuccess?.accept(it)

        }, {
            onError?.accept(if (it.message.equals(ERROR_NO_RESULT)) ERROR_NO_RESULT else ERROR_OTHERS)

        })) ?: onError?.accept(ERROR_OTHERS)
    }

    fun clearSearchResults() {
        mCurrentSearchResults.clear()
    }

    private fun setNextSearchLink(nextLink: String?, currentQuery: Map<String, String>) {
        if (nextLink == null) mHandle.set(KEY_LINK, null)
        else {
            val nextQuery = HashMap(currentQuery);
            nextQuery["page"] = getDigits(nextLink)

            mHandle.set(KEY_LINK, nextQuery)
        }
    }

    private fun getDigits(string: String): String {
        val num = StringBuilder()
        for (c in string.toCharArray()) {
            if (Character.isDigit(c)) num.append(c)
        }

        return num.toString()
    }

    fun getGenres(): List<String> {
        return mSearch.genre.map { it.dispName }
    }

    fun setGenres(genre: String, isAdded: Boolean) {
        val gen = Genre.values().find { it.dispName == genre }
        if (isAdded) mSearch.addGenre(gen)
        else mSearch.genre.remove(gen)
    }

    fun resetValues() {
        mSearch = MangaSearch()
        mHandle.keys().forEach { mHandle[it] = null }
    }

    fun search(): Map<String, String> {
        return mSearch.searchQuery()
    }

    fun createAndBookmarkManga(mangaLinks: ArrayList<String>, onError: () -> Unit) {
        Observable.fromIterable(mangaLinks).flatMap {
            mRepo.moveTo(it)
                    .subscribeOn(Schedulers.io())
                    .toManga(it)
                    .onErrorComplete {
                        onError.invoke()
                        true
                    }.doOnNext { it.isBookmark = true }
        }
                .toList()
                .subscribe { mangas -> mRepo.insertManga(*mangas.toTypedArray()) }
    }

    fun saveSearchParams() {
        val query = mSearch.searchQuery()
        query.keys.forEach {
            mHandle.set(it, query[it])
        }
    }

    fun createSearchFromHandle() {
        val lastQuery = hashMapOf<String, String>()
        mHandle.keys().forEach { if (!it.equals(KEY_LINK)) lastQuery[it] = mHandle[it]!! }

        mSearch = MangaSearch.from(lastQuery)
    }

    override fun onCleared() {
        super.onCleared()
        mCurrentSearchResults.clear()
    }
}
